Hazel/src/Hazel/Core/Input.h

```diff
@@ -14,12 +36,39 @@ namespace Hazel {
 		Input(const Input&) = delete;
 		Input& operator=(const Input&) = delete;
 
+		template <typename T>
+		static bool IsInputPressed(T name)
+		{
+			if (s_Instance->m_Mapping.count(static_cast<uint16_t>(name)) != 0)
+				return IsInputPressed(s_Instance->m_Mapping[static_cast<uint16_t>(name)]);
+
+			HZ_CORE_WARN("The input {0} is not mapped!", static_cast<uint16_t>(name));
+			return false;
+		}
+
+		inline static bool IsInputPressed(InputKey inputKey)
+		{
+			if (std::holds_alternative<KeyCode>(inputKey.value))
+				return s_Instance->IsKeyPressedImpl(std::get<KeyCode>(inputKey.value));
+
+			if (std::holds_alternative<MouseCode>(inputKey.value))
+				return s_Instance->IsMouseButtonPressedImpl(std::get<MouseCode>(inputKey.value));
+
+			// control should never fall down here;
+			HZ_CORE_WARN("Unknown Input type!");
+			return false;
+		}
+
 		inline static bool IsKeyPressed(KeyCode key) { return s_Instance->IsKeyPressedImpl(key); }
 
 		inline static bool IsMouseButtonPressed(MouseCode button) { return s_Instance->IsMouseButtonPressedImpl(button); }
 		inline static std::pair<float, float> GetMousePosition() { return s_Instance->GetMousePositionImpl(); }
 		inline static float GetMouseX() { return s_Instance->GetMouseXImpl(); }
 		inline static float GetMouseY() { return s_Instance->GetMouseYImpl(); }
+
+		/* Map a Key/Mouse Button to an InputKey, and overwrite the old value if existed. */
+		template <typename T>
+		static void Remap(T name, InputKey inputKey) { s_Instance->m_Mapping[static_cast<uint16_t>(name)] = inputKey; }
```

Comment:
Why not simply `static void Remap(InputName  name, InputKey inputKey)` ?

Reply:
The user may want to use their own custom input name, for example `InputName::JumpAndShoot` for a Right Click, which is not existed in the engine's `InputName` enum. If the function only accept `InputName` parameter then the user has to modify the engine code, which is what I want to avoid.
By making it a template function like that, user can define their own enum as long as its underlying type match the engine's one, which is `uint16_t` currently.

Line 70 should be `template <typename T : uint16_t>`.